pex/environment.py

```diff
@@ -27,19 +31,73 @@
 
 
 class PEXEnvironment(Environment):
+  class _CachingZipImporter(object):
+    class _CachingLoader(object):
+      def __init__(self, delegate):
+        self._delegate = delegate
+
+      def load_module(self, fullname):
+        loaded = sys.modules.get(fullname)
+        # Technically a PEP-302 loader should re-load the existing module object here - notably
+        # re-exec'ing the code found in the zip against the existing module __dict__. We don't do
+        # this since the zip is assumed immutable during our run and this is enough to work around
+        # the issue.
+        if not loaded:
+          loaded = self._delegate.load_module(fullname)
+          loaded.__loader__ = self
+        return loaded
+
+    _REGISTERED = False
+
+    @classmethod
+    def _ensure_namespace_handler_registered(cls):
+      if not cls._REGISTERED:
+        import pkg_resources
+        pkg_resources.register_namespace_handler(cls, pkg_resources.file_ns_handler)
+        cls._REGISTERED = True
+
+    def __init__(self, path):
+      import zipimport
+      self._delegate = zipimport.zipimporter(path)
+
+    def find_module(self, fullname, path=None):
+      loader = self._delegate.find_module(fullname, path)
+      if loader is None:
+        return None
+      self._ensure_namespace_handler_registered()
+      caching_loader = self._CachingLoader(loader)
+      return caching_loader
+
+  @classmethod
+  def _install_pypy_zipimporter_workaround(cls, pex_file):
+    def pypy_zipimporter_workaround(path):
+      import os
+
+      if not path.startswith(pex_file) or '.' in os.path.relpath(path, pex_file):
+        # We only need to claim the pex zipfile root modules.
+        raise ImportError()
```

Comment:
If PEX users could ever see this error, would be helpful to add a brief error message here.

Reply:
They can't, this is PEP-302 path hook import protocol - raise ImportError if your hook doesn't handle the given path; otherwise, return an importer. That said - it caused you questions, so added a comment with pep link.